### PR TITLE
Add TEP-0057: Windows Support

### PR DESCRIPTION
**Pull request by DrWadsy**
This TEP aims to address the need for Windows support in Tekton.

This was originally raised in [#1826](https://github.com/tektoncd/pipeline/issues/1826)

### CODE REVIEW

File: teps/0057-windows-support.md
Comment (ImJasonH) [inline teps/0057-windows-support.md:103]: Thanks for pointing these out. They do make me quite nervous 😄 

Can you think of any mitigations? We could explicitly state that while Tekton might _work_ on Windows, but that it isn't explicitly _supported_ for now, and until we get a bit more maturity we don't recommend it for real-world usage?

**Review by ImJasonH — COMMENTED** [review]
Have you done any experiments to determine what's broken today when you run a TaskRun scheduled to run on a Windows node? It'd be great to get an idea roughly how much work this will take to land.

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:93]: Sorry for the slow response, we've been pretty slammed this past fortnight.

We haven't been able to test this at all yet, we're still working on setting up a cluster with both windows and linux nodes so we can run real tests. While we've been using Tekton for a while now, we're still very new to the development side of things. I suspect the first thing, as you suggest, is modifying the entrypoint binary to be sure it correctly signals steps. Once this is done we'd actually want to attempt to run a TaskRun on a windows node, and see exactly what breaks.

I'd be very happy if it was as simple as `filepath.Join` but I'm not sure I'm that optimistic :)

Realistically we'll need some guidance from the Tekton team here, to point us towards the key, core pieces that will need work.

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:103]: Unfortunately the security issues around windows containers are not going to go away until a real fix for windows containers is implemented. Until then I think we should see use of windows containers as a 'here be dragons' situation - be careful using them, and absolutely don't run any untrusted workloads in them.

I'd agree that we should avoid recommending it for real-world usage until there is a decent degree of maturity present.

File: teps/0057-windows-support.md
Comment (ImJasonH) [reply teps/0057-windows-support.md:93]: Thanks @DrWadsy, that's helpful to know.

It sounds like we're still pretty far from even being able to identify exactly what work we'll need to do to get Windows support working. In that case, let's get the TEP submitted in roughly this form, and iterate on it with some experiments and prototyping. 

Would it be helpful to schedule some time where we can chat over video and better understand what the next steps would be? If so I'd be willing to find a time to do that.

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:51]: i think it makes sense to include these are requirements; i would say that (ideally) these components should support the Tasks which require Windows nodes _without_ requiring any changes (i.e. these components should all be decoupled enough that they don't need updates in order to support this)

possible exceptions:
* maybe the operator, if it needs to be aware of when Tasks that support windows should be installed (not sure it does any Task installation tho)
* if we were to update more components to run on windows (i.e. controllers, cli, dashboard) but that's outside the scope of this TEP anyway

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:66]: another possible non-goal to include is supporting PipelineResources (at least initially) - many of the images we build and publish as part of tekton pipelines (https://github.com/tektoncd/pipeline/tree/main/cmd) are just to support PipelineResources; I think it would be reasonable to not support these in windows tasks which would reduce the number of images we need to update (but maybe you need them anyway?)

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:89]: in the API working group yesterday @jerop pointed out that there is an issue suggesting this more generally as well: https://github.com/tektoncd/catalog/issues/661 - i think it makes sense to include this as a requirement, but i also could see hashing out the details of how we indicate the windows requirement in a separate TEP (i.e. so you don't have to tackle everything at once with this TEP!)

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:96]: @sbwsg pointed out in the [API working group notes](https://docs.google.com/document/d/17PodAxG8hV351fBhSu7Y_OIPhGTVgj6OJ2lPphYYRpU/edit#) that [pkg/pod/pod.go](https://github.com/tektoncd/pipeline/blob/main/pkg/pod/pod.go) might be a hotspot for windows related changes

File: teps/0057-windows-support.md
Comment (bobcatfish) [reply teps/0057-windows-support.md:103]: are there specific vulnerabilities we're concerned about?

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:106]: i dont know much about nodeSelectors and im wondering maybe you could provide an example - im wondering if the "nodeSelector" for windows nodes is always going to look the same, or if it could vary (i.e. will some clusters refer to their windows nodes as "win" and some refer to them as "windows", etc.) - because this user experience would suggest that the tekton pipelines controller needs to be able to look at the node selector and understand when it refers to a windows node (so it uses the appropriate images)

also afaik the nodeselector would likely be supplied at runtime (via a [pod template](https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/docs/podtemplates.md) in a TaskRun/Pipeline) - it seems like Tasks should have something in the Task spec that indicates if they are meant to be run on windows nodes or not

(let me know if this isnt making any sense - im not sure im explaining it well - im trying to refer to the difference between [runtime and authoring time](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability) - this feels like an authoring time thing - which leads me to the conclusion that the nodeselector is maybe not the way the controller detects that it needs to use windows compatible images)

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:128]: i mentioned this in a comment above, but im wondering it makes more sense for this info to be in the Task vs the TaskRun (though the TaskRun might ultimately need to provide a nodeSelector as well 😬 )

File: teps/0057-windows-support.md
Comment (bobcatfish) [inline teps/0057-windows-support.md:138]: we're also going to need (in the long run) a way to test on Windows - which means we'll need to add that to our existing test infra (which is totally reasonable! but worth calling out i think :D)

**Review by bobcatfish — COMMENTED** [review]
Looking great! I have some questions around Task vs TaskRun changes but generally I feel like this would be great to merge as an initial problem statement and then start iterating on the design :D

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:51]: I think that's definitely what we'd be aiming for - current elements should keep working as is, with no changes necessary, other thank the elements which need to handle the Windows nodes. Everything else should just keep ticking as it already is.

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:66]: That's a fair point - we don't need PipelineResources for our use case, so if we're happy to leave this until later that works for us :)

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:89]: Agreed - the mechanics of this can probably wait, we just wanted to indicate that if a Task will require a Windows node then this should be flagged as early as possible for user-friendliness.

File: teps/0057-windows-support.md
Comment (DrWadsy) [reply teps/0057-windows-support.md:106]: I can't respond to the comment above, so I'll answer that here first. The big one is that it is possible to break-out of containers when using Windows (one example is documented [here](https://unit42.paloaltonetworks.com/windows-server-containers-vulnerabilities/). This is something that requires a fix from Microsoft, there's nothing that can be done at our end, bit obviously this is something we need to be conscious of moving forward.

Comment (DrWadsy) [reply teps/0057-windows-support.md:106]: In terms of nodeSelectors - as I understand it kubernetes provides a fairly well defined way to require that a container be scheduled on a node that meets certain requirements. These can include OS requirements as described [here](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/). I'd agree that we'd probably supply the nodeSelector at runtime, but we would definitely want something in the Task (i.e. authoring time)  that indicates this need. So maybe we do need something other than a nodeSelector here, but I'd imagine we'd want it to be defined along similar lines...

Another issue we need to face is the lack of flexibility when using windows containers on windows nodes - unless things have changed very recently then Windows _requires_ that the windows version of the node and container be the same.

Comment (bobcatfish) [reply teps/0057-windows-support.md:106]: re. the vulnerabilites: thanks for the example! might be worth including that link in the content of the TEP as well

re. nodeSelectors:

Thanks for the explanation, I think that makes sense to me. My conclusion is that we ultimately will need to have something in the Task that indicates this need, and then at runtime folks will need to include the appropriate nodeselector for their cluster as well but (correct me if I'm wrong!!) Tekton probably can't be too opinionated about what the node selector looks like (since folks could be using any label they want to identify their windows nodes)? Or maybe there are already known conventions around identifying windows nodes we could embrace

**Comment by bobcatfish** [comment]
I'm happy to merge this TEP as is and keep iterating on it!

/approve

@imjasonh @afrittoli  do you have any blockers for this PR?

**Review by vdemeester — APPROVED** [review]
Agreed with @bobcatfish

**Comment by sbwsg** [comment]
Adding and releasing a hold to see if it unblocks the merge queue

/hold

**Comment by sbwsg** [comment]
/hold cancel

It looks like this PR might need to be rebased.  Or at least that's what tide is implying with [the error messages in its history](https://prow.tekton.dev/tide-history):

> failed merging [383]: [PR is unmergable. Do the Tide merge requirements match the GitHub settings for the repo? This branch can't be rebased]

**Comment by sbwsg** [comment]
Just a quick ping - rebase on latest `main` branch and this should get automatically merged.

**Comment by AlanGreene** [comment]
The label will be removed automatically anyway when the PR is rebased, removing it manually to see if it unblocks things / stops spamming Tide.
/lgtm cancel

**Comment by DrWadsy** [comment]
Sorry, I was away for a few days. Rebased this morning, looks like it's just waiting on that lgtm label :)

**Pull request merged**
